src/libopensc/card-sc-hsm.c

```diff
@@ -1883,14 +1910,18 @@ static int sc_hsm_finish(sc_card_t * card)
 #ifdef ENABLE_SM
 	sc_sm_stop(card);
 #endif
-	if (priv->serialno) {
-		free(priv->serialno);
-	}
-	if (priv->dffcp) {
-		sc_file_free(priv->dffcp);
+	if (priv) {
+		if (priv->serialno) {
+			free(priv->serialno);
+		}
+		if (priv->dffcp) {
+			sc_file_free(priv->dffcp);
+		}
+		if (priv->EF_C_DevAut) {
+			free(priv->EF_C_DevAut);
+		}
```

Comment:
the null check like this is useless as `free` handles NULL arguments without any issues and when the member already contains some junk data, it will not save you.

Reply:
Yes. This was already in place before this PR.

There are lots of other occurrences in the code, esp. when calling `sc_file_free`. I will remove these spurious checks as well. However, the added `if (priv) {` check prevents to dereference a `NULL` value of `priv` in `free(priv->serialno);`.